lib/handlers/pools/provider-migration/v3/traffic-switch-pool-provider.ts

```diff
@@ -0,0 +1,103 @@
+import { Token } from '@uniswap/sdk-core'
+import { FeeAmount, Pool } from '@uniswap/v3-sdk'
+import { ProviderConfig } from '@uniswap/smart-order-router/build/main/providers/provider'
+import { POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION } from './pool-provider-traffic-switch-configuration'
+
+export type TrafficSwitchPoolProviderProps = {
+  currentPoolProvider: IV3PoolProvider
+  targetPoolProvider: IV3PoolProvider
+  sourceOfTruthPoolProvider: IV3PoolProvider
+}
+
+export class TrafficSwitchPoolProvider implements IV3PoolProvider {
+  private readonly currentPoolProvider: IV3PoolProvider
+  private readonly targetPoolProvider: IV3PoolProvider
+  private readonly sourceOfTruthPoolProvider: IV3PoolProvider
+
+  private readonly SHOULD_SWITCH_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.switchPercentage > this.getRandomPercentage()
+  private readonly SHOULD_SAMPLE_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.samplingPercentage > this.getRandomPercentage()
+
+  constructor({ currentPoolProvider, targetPoolProvider, sourceOfTruthPoolProvider }: TrafficSwitchPoolProviderProps) {
+    this.currentPoolProvider = currentPoolProvider
+    this.targetPoolProvider = targetPoolProvider
+    this.sourceOfTruthPoolProvider = sourceOfTruthPoolProvider
+  }
+
+  getPoolAddress(
+    tokenA: Token,
+    tokenB: Token,
+    feeAmount: FeeAmount
+  ): {
+    poolAddress: string
+    token0: Token
+    token1: Token
+  } {
+    // The underlying logic for getting the pool address is always from the
+    // source of truth pool provider, and the result is deterministic.
+    // It doesn't matter which pool provider we use, so we can start with the target
+    // pool provider, which will delegate to the source of truth pool provider.
+    return this.targetPoolProvider.getPoolAddress(tokenA, tokenB, feeAmount)
+  }
+
+  async getPools(tokenPairs: [Token, Token, FeeAmount][], providerConfig?: ProviderConfig): Promise<V3PoolAccessor> {
+    const currentProviderPools = await this.currentPoolProvider.getPools(tokenPairs, providerConfig)
+    const targetProviderPools = await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+
+    const sampleTraffic = this.SHOULD_SAMPLE_TRAFFIC()
+    if (sampleTraffic) {
+      // If we need to sample the traffic, we don't want to make it a blocking I/O
+      this.sampleTraffic(tokenPairs, currentProviderPools, targetProviderPools, providerConfig)
+    }
+
+    const switchTraffic = this.SHOULD_SWITCH_TRAFFIC()
+    if (switchTraffic) {
+      return currentProviderPools
+    } else {
+      return targetProviderPools
+    }
+  }
+
+  private async sampleTraffic(
+    tokenPairs: [Token, Token, FeeAmount][],
+    currentProviderPools: V3PoolAccessor,
+    targetProviderPools: V3PoolAccessor,
+    providerConfig?: ProviderConfig
+  ): Promise<void> {
+    const truthProviderPools = await this.sourceOfTruthPoolProvider.getPools(tokenPairs, providerConfig)
```

Comment:
I get what we are doing here, comparing the old and new methods against the actual sourceOfTruth... this could slow down the sampled requests, right?